File: compiler/src/pipeline/steps/type_registration.py
"""Type registration step implementation."""

from dataclasses import replace
from .base import MacroProcessingStep
from core.macro_registry import MacroContext, MacroRegistry
from core.node import Macro


class TypeRegistrationStep(MacroProcessingStep):
    """Handles the first pass of type checking: registering types and functions."""
    
    def __init__(self, macros: MacroRegistry):
        super().__init__()
        self.macros = macros
        
    def process_node(self, ctx: MacroContext) -> None:
        macro = str(ctx.compiler.get_metadata(ctx.node, Macro))
        all_macros = self.macros.all()

        if macro in all_macros:
            # Call the register_type method on the macro provider
            all_macros[macro](ctx)

        # Recursively process children to find nested definitions
        for child in ctx.node.children:
            child_ctx = replace(ctx, node=child)
            self.process_node(child_ctx)


class TypeDetailRegistrationStep(MacroProcessingStep):
    """Handles the second pass of type registration: filling in field details with proper types."""
    
    def __init__(self, macros: MacroRegistry):
        super().__init__()
        self.macros = macros
        
    def process_node(self, ctx: MacroContext) -> None:
        macro = str(ctx.compiler.get_metadata(ctx.node, Macro))
        all_macros = self.macros.all()

        if macro in all_macros:
            # Call the registered function directly
            all_macros[macro](ctx)

        # Recursively process children
        for child in ctx.node.children:
            child_ctx = replace(ctx, node=child)
            self.process_node(child_ctx)

File: compiler/src_ts/pipeline/steps/type_registration.ts
// pipeline/steps/type_registration.ts

import { MacroProcessingStep } from "./base.ts";
import { MacroContext, MacroRegistry, TCResult } from "../../core/macro_registry.ts";
import { Macro } from "../../core/node.ts";

export class TypeRegistrationStep extends MacroProcessingStep {
  constructor(public override macros: MacroRegistry) {
    super();
    this.macros = macros;
  }

  process_node(ctx: MacroContext): null {
    const macroName = String(ctx.compiler.get_metadata(ctx.node, Macro));
    const all = this.macros.all();

    if (Object.prototype.hasOwnProperty.call(all, macroName)) {
      all[macroName].call(ctx, ctx);
    }

    ctx.node.children.forEach((child) => {
      const childCtx = ctx.clone_with({ node: child });
      this.process_node(childCtx);
    });

    return null;
  }
}

export class TypeDetailRegistrationStep extends MacroProcessingStep {
  constructor(public override macros: MacroRegistry) {
    super();
    this.macros = macros;
  }

  process_node(ctx: MacroContext): null {
    const macroName = String(ctx.compiler.get_metadata(ctx.node, Macro));
    const all = this.macros.all();

    if (Object.prototype.hasOwnProperty.call(all, macroName)) {
      all[macroName].call(ctx, ctx);
    }

    ctx.node.children.forEach((child) => {
      const childCtx = ctx.clone_with({ node: child });
      this.process_node(childCtx);
    });

    return null;
  }
}

export class FunctionRegistrationStep extends MacroProcessingStep {
  constructor(public override macros: MacroRegistry) {
    super();
    this.macros = macros;
  }

  process_node(ctx: MacroContext): TCResult {
    const macroName = String(ctx.compiler.get_metadata(ctx.node, Macro));
    const all = this.macros.all();

    if (Object.prototype.hasOwnProperty.call(all, macroName)) {
      return all[macroName].call(ctx, ctx);
    }

    ctx.node.children.forEach((child) => {
      const childCtx = ctx.clone_with({ node: child });
      this.process_node(childCtx);
    });

    return null;
  }
}
